File: src/services/team/contextManager.ts
import { LLMProvider } from './llm';

const CHARS_PER_TOKEN = 3;

// Layer 1
const TOOL_PROTECTION_THRESHOLD = 30000;
const MIN_PRUNABLE_THRESHOLD = 15000;
const MAX_TOOL_RESULT_CHARS = 8000;
const PREVIEW_HEAD_CHARS = 500;
const PREVIEW_TAIL_CHARS = 500;

// Layer 2
const COMPRESSION_TOKEN_THRESHOLD = 0.45;
const COMPRESSION_PRESERVE_RATIO = 0.35;

// Layer 3
const OVERFLOW_SAFETY_MARGIN = 0.85;

export function estimateTokens(text: string): number {
    if (!text) return 0;
    return Math.ceil(text.length / CHARS_PER_TOKEN);
}

export function estimateMessageTokens(msg: any): number {
    let text = msg.content || '';
    if (msg.tool_calls) {
        text += JSON.stringify(msg.tool_calls);
    }
    return estimateTokens(text);
}

export function estimateMessagesTokens(messages: any[]): number {
    return messages.reduce((acc, msg) => acc + estimateMessageTokens(msg), 0);
}

// ----------------------------------------------------------------------------
// Layer 1: Tool Output Masking
// ----------------------------------------------------------------------------
export function maskToolOutputs(messages: any[]): any[] {
    let accumulatedToolTokens = 0;
    let prunableTokens = 0;
    const itemsToPrune: any[] = [];
    const newMessages = [...messages];

    // Reverse scan to protect recent tool outputs
    for (let i = newMessages.length - 1; i >= 0; i--) {
        const msg = newMessages[i];
        if (msg.role === 'tool') {
            const contentStr = typeof msg.content === 'string' ? msg.content : JSON.stringify(msg.content);
            const tokens = estimateTokens(contentStr);

            if (accumulatedToolTokens < TOOL_PROTECTION_THRESHOLD) {
                // Inside protection window
                accumulatedToolTokens += tokens;
            } else {
                // Outside protection window
                if (contentStr.length > MAX_TOOL_RESULT_CHARS) {
                    prunableTokens += tokens;
                    itemsToPrune.push({ index: i, originalContent: contentStr });
                }
            }
        }
    }

    // Only execute if batch threshold is met
    if (prunableTokens >= MIN_PRUNABLE_THRESHOLD) {
        for (const item of itemsToPrune) {
            const originalContent: string = item.originalContent;
            const head = originalContent.substring(0, PREVIEW_HEAD_CHARS);
            const tail = originalContent.substring(originalContent.length - PREVIEW_TAIL_CHARS);

            const omittedLines = originalContent.substring(PREVIEW_HEAD_CHARS, originalContent.length - PREVIEW_TAIL_CHARS).split('\n').length;
            const approxKB = Math.round(originalContent.length / 1024);

            newMessages[item.index].content = `[Tool output truncated — original: ~${omittedLines} lines, ~${approxKB}KB]\n${head}\n...\n[${omittedLines} lines omitted]\n...\n${tail}`;
        }
        console.log(`[ContextManager] Masked ${itemsToPrune.length} tool outputs, saved ~${prunableTokens} tokens.`);
    }

    return newMessages;
}

// ----------------------------------------------------------------------------
// Layer 2: Chat Compression
// ----------------------------------------------------------------------------
export async function compressChat(messages: any[], llm: LLMProvider, contextLimit: number = 131072): Promise<{ compressed: boolean, messages: any[] }> {
    const currentTokens = estimateMessagesTokens(messages);
    const threshold = contextLimit * COMPRESSION_TOKEN_THRESHOLD;

    if (currentTokens <= threshold) {
        return { compressed: false, messages };
    }

    // Attempt compression
    // Split messages at approx 65% / 35% ratio
    const preserveRatioTokens = currentTokens * COMPRESSION_PRESERVE_RATIO;

    let splitIndex = messages.length - 1;
    let accumulatedFromEnd = 0;

    // Find the split point going backwards to ensure we preserve the latest chunks
    while (splitIndex >= 0 && accumulatedFromEnd < preserveRatioTokens) {
        accumulatedFromEnd += estimateMessageTokens(messages[splitIndex]);
        splitIndex--;
    }

    // Ensure split index lands on a 'user' message to not break tool_call -> tool_result pairs
    while (splitIndex >= 0 && messages[splitIndex].role !== 'user') {
        splitIndex--;
    }

    if (splitIndex <= 0) {
        // Cannot split safely
        return { compressed: false, messages };
    }

    const messagesToCompress = messages.slice(0, splitIndex);
    const messagesToPreserve = messages.slice(splitIndex);

    const prompt = `You are a context compression assistant. Summarize the following dialogue history.
Rules:
1. Retain all file paths, function names, variable names, error messages, tool results, and technical decisions.
2. Retain user preferences, constraints, and explicit requirements.
3. Record which tools were used and key findings.
4. Track the current state of ongoing tasks.
5. Be concise but preserve executable context.
6. Output in the language of the conversation.
7. Wrap your output in <state_snapshot> tags.

History to compress:
${JSON.stringify(messagesToCompress, null, 2)}`;

    try {
        const response = await llm.chat([{ role: 'user', content: prompt }]);
        const snapshot = response.content || '';

        const newMessages = [
            {
                role: 'user',
                content: `[Context Snapshot — compressed from ${messagesToCompress.length} messages]\n\n${snapshot}`
            },
            {
                role: 'assistant',
                content: '明白，我已掌握之前的上下文。请继续。'
            },
            ...messagesToPreserve
        ];

        const afterTokens = estimateMessagesTokens(newMessages);

        // Deflation check
        if (afterTokens >= currentTokens) {
            console.warn('[ContextManager] Compression inflated context, reverting.');
            return { compressed: false, messages };
        }

        console.log(`[ContextManager] Compressed History: ${currentTokens} -> ${afterTokens} tokens.`);
        return { compressed: true, messages: newMessages };
    } catch (e) {
        console.error('[ContextManager] Compression failed', e);
        return { compressed: false, messages };
    }
}

// ----------------------------------------------------------------------------
// Layer 3: Overflow Prevention
// ----------------------------------------------------------------------------
export function checkOverflow(messages: any[], systemPrompt: string, contextLimit: number = 131072): { safe: boolean, totalTokens: number, limit: number, usagePercent: number } {
    const sysTokens = estimateTokens(systemPrompt);
    const msgTokens = estimateMessagesTokens(messages);
    const totalTokens = sysTokens + msgTokens;

    const hardLimit = contextLimit * OVERFLOW_SAFETY_MARGIN;
    const usagePercent = Math.round((totalTokens / contextLimit) * 100);

    return {
        safe: totalTokens < hardLimit,
        totalTokens,
        limit: contextLimit,
        usagePercent
    };
}

File: src-tauri/src/modules/agent/context_manager.rs
use agents_sdk::messaging::{AgentMessage, MessageContent, MessageRole};
use serde_json::Value;
use tracing::{info, warn};

// Estimations
const CHARS_PER_TOKEN: usize = 3;

// Layer 1
const TOOL_PROTECTION_THRESHOLD: usize = 30000;
const MIN_PRUNABLE_THRESHOLD: usize = 15000;
const MAX_TOOL_RESULT_CHARS: usize = 8000;
const PREVIEW_HEAD_CHARS: usize = 500;
const PREVIEW_TAIL_CHARS: usize = 500;

// Layer 2
const COMPRESSION_TOKEN_THRESHOLD: f64 = 0.45;
const COMPRESSION_PRESERVE_RATIO: f64 = 0.35;

// Layer 3
const OVERFLOW_SAFETY_MARGIN: f64 = 0.85;

pub fn estimate_tokens(text: &str) -> usize {
    if text.is_empty() {
        return 0;
    }
    (text.len() as f64 / CHARS_PER_TOKEN as f64).ceil() as usize
}

pub fn estimate_message_tokens(msg: &AgentMessage) -> usize {
    let mut text_len = 0;
    match &msg.content {
        MessageContent::Text(t) => text_len += t.len(),
        MessageContent::Json(v) => text_len += v.to_string().len(),
    }
    (text_len as f64 / CHARS_PER_TOKEN as f64).ceil() as usize
}

pub fn estimate_messages_tokens(messages: &[AgentMessage]) -> usize {
    messages.iter().map(estimate_message_tokens).sum()
}

// ----------------------------------------------------------------------------
// Layer 1: Tool Output Masking
// ----------------------------------------------------------------------------
pub fn mask_tool_outputs(messages: &[AgentMessage]) -> Vec<AgentMessage> {
    let mut accumulated_tool_tokens = 0;
    let mut prunable_tokens = 0;
    
    // Track pairs of (index, original_text, tokens)
    let mut items_to_prune: Vec<(usize, String, usize)> = Vec::new();
    let mut new_messages = messages.to_vec();

    // Reverse scan
    for (i, msg) in new_messages.iter().enumerate().rev() {
        if msg.role == MessageRole::Tool {
            let content_str = match &msg.content {
                MessageContent::Text(t) => t.clone(),
                MessageContent::Json(v) => v.to_string(),
            };
            
            let tokens = estimate_tokens(&content_str);

            if accumulated_tool_tokens < TOOL_PROTECTION_THRESHOLD {
                accumulated_tool_tokens += tokens;
            } else if content_str.len() > MAX_TOOL_RESULT_CHARS {
                prunable_tokens += tokens;
                items_to_prune.push((i, content_str.clone(), tokens));
            }
        }
    }

    if prunable_tokens >= MIN_PRUNABLE_THRESHOLD {
        for (idx, original_content, _) in items_to_prune.iter() {
            let head: String = original_content.chars().take(PREVIEW_HEAD_CHARS).collect();
            let tail_start = original_content.chars().count().saturating_sub(PREVIEW_TAIL_CHARS);
            let tail: String = original_content.chars().skip(tail_start).collect();
            
            let omitted_lines = original_content.lines().count().saturating_sub(
                head.lines().count() + tail.lines().count()
            );
            
            let approx_kb = original_content.len() / 1024;
            
            let replacement = format!(
                "[Tool output truncated — original: ~{} lines, ~{}KB]\n{}\n...\n[{} lines omitted]\n...\n{}",
                omitted_lines, approx_kb, head, omitted_lines, tail
            );
            
            new_messages[*idx].content = MessageContent::Text(replacement);
        }
        
        info!("[ContextManager] Masked {} tool outputs, saved ~{} tokens.", items_to_prune.len(), prunable_tokens);
    }

    new_messages
}

// ----------------------------------------------------------------------------
// Layer 2: Chat Compression (LLM call omitted for wrapper simplicity, could be added later if needed)
// ----------------------------------------------------------------------------
// Note: Chat compaction is already handled in the background by memory.rs in Helix.
// The true 3-Layer compression requires interrupting the SDK loop, 
// but for now, we will focus on Masking and Overflow clipping which are 100% reliable.

// ----------------------------------------------------------------------------
// Layer 3: Overflow Prevention
// ----------------------------------------------------------------------------
pub struct OverflowStatus {
    pub safe: bool,
    pub total_tokens: usize,
    pub limit: usize,
    pub usage_percent: usize,
}

pub fn check_overflow(messages: &[AgentMessage], context_limit: usize) -> OverflowStatus {
    // Assuming context Limit defaults to 131072 if not provided exactly
    let limit = if context_limit == 0 { 131072 } else { context_limit };
    let total_tokens = estimate_messages_tokens(messages);
    
    let hard_limit = (limit as f64 * OVERFLOW_SAFETY_MARGIN) as usize;
    let usage_percent = ((total_tokens as f64 / limit as f64) * 100.0).round() as usize;

    OverflowStatus {
        safe: total_tokens < hard_limit,
        total_tokens,
        limit,
        usage_percent,
    }
}

/// Helper method for InterceptingChatModel
pub fn emergency_trim(messages: &mut Vec<AgentMessage>) {
    if messages.len() > 10 {
        // Keep system prompt + the last 30% of messages
        let retain_len = (messages.len() as f64 * 0.3).ceil() as usize;
        let start_idx = messages.len() - retain_len;
        
        let mut trimmed = Vec::new();
        // Assume first might be System
        if !messages.is_empty() && messages[0].role == MessageRole::System {
            trimmed.push(messages[0].clone());
        }
        
            trimmed.push(AgentMessage {
                role: MessageRole::System,
                content: MessageContent::Text("[Earlier context was truncated due to context window overflow]".to_string()),
                metadata: None,
            });
        
        let slice = &messages.clone()[start_idx..];
        for m in slice {
            trimmed.push(m.clone());
        }
        
        *messages = trimmed;
        warn!("[ContextManager] Emergency trim applied due to overflow.");
    }
}
